### ISSUE TITLE
Regression Testing tool for javadoc parsing.

### ISSUE THREAD

**Issue by ps-sp** [checkstyle/contribution#237]
UTs for javadocs are insufficient. Also they currently have a poor implementation style ([4381](https://github.com/checkstyle/checkstyle/issues/4381)).

Thus, regression testing is a must for most of the grammar updates. The parse trees in the master and the patch should be compared and the differences should be reported.

**Comment by Vladlis** [comment]
The initial thought was to make a simple script to do comparison of CLI output, but results would have been not representative and hard to analyze.

[Proposal](https://github.com/checkstyle/checkstyle/pull/4378#issuecomment-304561217) from @rnveach is to create a utility check that would only print AST as a violation so that we could use the existing checkstyle-tester for reports. I think this makes more sense even though
> the XML files (temporary report files) may be larger

For me it is not clear how are we going to use such a check:
- locally add it to the project every time (unlikely)
- add it as a utility to the main project?
- add like a third party check?

@romani @rnveach @sabaka Please join the conversation.

**Comment by rnveach** [comment]
> simple script to do comparison of CLI output

This was my second plan, if the first didn't work out.
One issue is the main goal was to compare the Javadoc structure. To use existing projects for regression, we need to first identify the Javadocs in the files. The CLI currently can't do this unless we use the `-J` option and cut out the non-Javadoc syntax, which would require a completely new process.
Another issue is regression needs to be able to show the original Java file in the web so we know what the differences are we are looking at. CLI doesn't do this, so we would need to apply JXR from normal regression to this new process somehow.
We would also need a utility to find similarities and differences, even across multiple lines, as new nodes would cause existing lines to be moved down. patch-diff-report-util can't handle this type of comparison.

There are some issues with the way I implemented the check, which may also cause some issues for regression.
Line/column numbers won't match up to the actual line/column numbers in the file. I made a hack just to make sure the line numbers are unique, as we sort the results based on those and incorrect sorting will make the printed tree hard to review.
Also because of the above, new/removed lines in the differences will affect all lines after it because line numbers is based on AST output line number. It won't see the same text on different lines as the same violation. This would still be an issue with the CLI way too and I said with it.
Doing all the files in the project will create a lot of violations, and may make the final reports large depending on the number of changes. I saw a huge slowdown in the openjdks projects, when testing this out, so I recommend avoiding them.

> add like a third party check

This seems like the best option to me. We would have our own set of checks for testing regression.

**Comment by romani** [comment]
@ps-sp , @Vladlis , @Vladlis 

this issue is good start of big problem, the same problem we had at 2014 when we come to regression testing, so that is when very initial regression tool is started. It was partly automated, and in 3 years become very good application to ease testing by users who do not want spend too much time on testing of their PRs.

DO NOT dream to make the same tool right now! you do not see whole scope of problems and there are no time to make it (GSOC is limited by time and we have big goals). Ideas could be continue to be collected ....

We need for now very simple automation of regression in AST.
MAIN POINT: we do not expect regressions(changes) in most cases!!! it will simplify all a bit.

> The CLI currently can't do this unless we use the -J option and cut out the non-Javadoc syntax, which would require a completely new process.

it is enough for now. All we need is to save output in same path, same name but extension ".javadoc" it will ease to find source file if required.

> Another issue is regression needs to be able to show the original Java file in the web so we know what the differences are we are looking at.

web ready report is good dream, but lets focus on PRs that we have now. We do not need complete solution for users who unaware of regression details. This tool will be used by us, so it is targeted to advanced users with big context of what is going on.

> We would also need a utility to find similarities and differences, even across multiple lines

as we have two folders of `*.javadoc` files we can run meld - http://meldmerge.org/ . It will do all that we need to catch regresion points easily.
It might be even `diff` could be useful for some simple regressions.

So imagine we have script(bash or ....) that generate javadoc foder from all java files in source folder.
We run it for master and for patch. Meld helps to analyse it.
if regression is big, use less code to be able to anlyse it manually.
So author of fix/PR will do this himself of some code. if any concerns - raise hand and share folders to mentors to review and point to problematic cases. Author also share commands how he did generation to make it reproducible.
Mentors do verification in Meld.  Cases for discussion could be places to free web services for diff, there are bunch of them.

--------------------------------------

this process will help to see corner cases and eventually more steps will be automated and make convenient.

**Comment by ps-sp** [comment]
@romani So the final conclusion is to use a script rather than trying to [reuse](https://github.com/checkstyle/checkstyle/pull/4378#issuecomment-304561217) the existing regression tool using an adhoc check ?

**Comment by romani** [comment]
I still wait for @Vladlis  and @rnveach for feedback.
But I do not see how can they propose more cheap solution.

**Comment by rnveach** [comment]
My opinion, if we want fast, I don't think we can do faster than the custom check. Only issue is applying it to both branches as [it requires a small change in to a base class to method that is marked as `final` to be able to identify when a new file is being processed](https://github.com/rnveach/checkstyle/commit/6a2600ca9d779e17a5e9076aa0da3d0abf14f71f#diff-8582d74a6dba42fdf44e42ee5fc88127L243).

In the end, I think we do need a new process for the compare. 
The method @romani mentioned for the new process is fine.
I am looking over `patch-diff-report-util` to have it do the text file compare using JGit, not to wait on @ps-sp 's work or using meld for temporary.

**Comment by Vladlis** [comment]
I agree with @rnveach. Looks like a utility check will be faster and easier to use and review.

**Comment by ps-sp** [comment]
@romani Please address. This is a blocking issue for all grammar related issues.

**Comment by rnveach** [comment]
@ps-sp Until we come to a decision on a final approach, I recommend using the shell script I mentioned at https://github.com/checkstyle/checkstyle/pull/4386#issuecomment-305771090 which I used to generate reports for that PR.

**Comment by ps-sp** [comment]
@rnveach ok

**Comment by rnveach** [comment]
Some benefits for the Check over a bash script is the limitations of the CLI and bash itself.
The command line `-J` option can only do 1 file at a time. Bash is not very good at multi-threading. It can only run things in parallel. Even with this, running regression over the Checkstyle project alone takes 30 minutes on my machine. Once multi-threading is implemented in Checkstyle, a check will benefit from the speed improvements.

**Comment by rnveach** [comment]
This is the current version of my shell script, which is not a check. We will beta it out for PRs for now to see how it works out.

https://github.com/rnveach/contribution/commits/antlr_regression (see top 4 commits)
The script is `launch_diff_antlr.sh` and only takes the PR branch name as an input.
I recommend modifying `launch_diff_variables.sh` to your setup.
`patch-diff-report-tool` needs to be re-packaged because it adds some new, non-breaking changes to the reporting.

I don't recommend running this on the openjdk projects, as printing ANTLR is a slow task. Checkstyle project for me takes 30 minutes.

**Comment by romani** [comment]
@ps-sp , please use work of rnveach, I need to know that you can work with this tool yourself and you can provide regression testing yourself to speed up PR acceptance.

@rnveach , I see strange regression on incorrect recognition of LEADING_ASTERISK
http://rveach.no-ip.org/checkstyle/regression/reports/67/guava/index.html#A1 , it is not expected.
exceptions become less informative - http://rveach.no-ip.org/checkstyle/regression/reports/68/checkstyle/index.html#A1 , not a big deal but ..... very suspicious.

**Comment by rnveach** [comment]
@romani those were just test runs that I used to verify the script was working.
`67` is for one of ps-sp's PR. We are working through that issue.
`68` was me playing with grammar making some modifications to see what affect they had.

**Comment by romani** [comment]
ok, lets wait for @ps-sp results

**Comment by ps-sp** [comment]
@romani @rnveach I will try out rnveach scripts and get back to you. Meanwhile I made [this](https://github.com/ps-sp/contribution/commit/a746eb6b0fe4aa509d74556e376d527d89671c24#diff-1621fab066313a75e8614be6c1cab922) groovy script for regression. It is unbelievably slow. I was wondering why do we [NOT](https://github.com/checkstyle/checkstyle/blob/7ba18e5138cdb800243b0a7dd15de3bb3741e395/src/main/java/com/puppycrawl/tools/checkstyle/Main.java#L260) allow printing the javadoc tree directly to an output file ? Also we don't allow printing javadocs for multiple files in a folder. Were there issues with allowing this previously or is it just because printing javadocs is tedious ?

**Comment by romani** [comment]
> I will try out rnveach scripts and get back to you

I merged his PR.

>  I was wondering why do we NOT allow printing the javadoc tree directly to an output file ?

we keep our Main a bit clear, as we force users to use it, for there should be less obscure parameters that required in very specific cases. I doubt that performance problems is related to redirection of output to file. Performance process is due to slow parsing.

> Also we don't allow printing javadocs for multiple files in a folder.

we do not have this for java code too, so the same behavior is for javadoc. Printing of javadoc tree is required only for one time 99% of time, when user do debug. Only your script need massive generation of AST trees.

> Were there issues trying to support this previously or is it just because printing javadocs is tedious ?

it is not required is real life usage of Main , so no support in Main.

---------------------

@ps-sp , please run regression by @rnveach code. We need to merge PRs as soon as possible. 
If can prove that tour script is better than script of @rnveach , we could host it too.

**Comment by ps-sp** [comment]
@romani I am facing some issues with @rnveach scripts. I have already contacted him about it. My script is very slow, anyways I am trying to optimize things until I am able to get @rnveach scripts running.

**Comment by romani** [comment]
@ps-sp , please do not go crazy with optimization of it, run it overnight or when you have a break.
you need to focus on different optimization.

**Comment by rnveach** [comment]
@ps-sp Trust me, my script will not be any faster.
Like romani said, true slowdown is ANTLR parsing.

**Comment by ps-sp** [comment]
I meant I was optimizing things _with my script_. But yeah I get it, I won't waste my time with that.
@rnveach Did you look at the logs I sent you ?

**Comment by rnveach** [comment]
@ps-sp Yes, but I can't really reply well from work as I can't access GMail directly and working with my phone isn't always fun.
I suggest we move discussion to either gitter or google groups.

Are you using Windows to execute the bash script or Linux?
The last 2 lines of your output look garbled and issue is it's trying to work with an empty line as a repo. We have seen something similar only in appveyor when running shell scripts for windows based on new version of the executor.

**Comment by ps-sp** [comment]
@rnveach Yep, on windows.I will try it out on ubuntu then. We can talk further in this: https://groups.google.com/forum/?hl=en#!topic/checkstyle-devel/cRw80h4Q73o

**Comment by ps-sp** [comment]
@rnveach @romani works for me now. I will generate more diff reports and upload them as well. [Here](https://issue-2329-diff-report1.netlify.com/) is the diff report for Guava for #2329

**Comment by ps-sp** [comment]
We have a regression testing tool ready in form of shell script: [launch_diff_antlr.sh](https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch_diff_antlr.sh). Hence, I am closing this issue.
